tecken/tests/test_upload.py

```diff
@@ -542,314 +312,155 @@ def mock_api_call(self, operation_name, api_params):
 def test_upload_archive_key_lookup_cached(
     client,
-    botomock,
-    fakeuser,
+    db,
+    s3_helper,
+    uploaderuser,
 ):
-    token = Token.objects.create(user=fakeuser)
+    token = Token.objects.create(user=uploaderuser)
     (permission,) = Permission.objects.filter(codename="upload_symbols")
     token.permissions.add(permission)
-    url = reverse("upload:upload_archive")
-
-    lookups = []
+    s3_helper.create_bucket("publicbucket")
 
-    metadata_cache = {}
-
-    def mock_api_call(self, operation_name, api_params):
-        # This comes for the setting UPLOAD_DEFAULT_URL specifically
-        # for tests.
-        assert api_params["Bucket"] == "publicbucket"
-        if operation_name == "HeadBucket":
-            # yep, bucket exists
-            return {}
-
-        if operation_name == "HeadObject" and api_params["Key"] == (
-            "v0/flag/deadbeef/flag.jpeg"
-        ):
-            return {"ContentLength": 69183}
-
-        if operation_name == "HeadObject" and api_params["Key"] == (
-            "v0/xpcshell.dbg/A7D6F1BB18CD4CB48/xpcshell.sym"
-        ):
-            # Saying the size is 100 will cause the code to think the
-            # symbol file *is different* so it'll proceed to upload it.
-            size = 100
-            if lookups:
-                # If this is the second time, return the right size.
-                size = 501
-            result = {"ContentLength": size}
-            if metadata_cache.get(api_params["Key"]):
-                result["Metadata"] = metadata_cache[api_params["Key"]]
-            lookups.append(size)
-            return result
-
-        if operation_name == "PutObject" and api_params["Key"] == (
-            "v0/xpcshell.dbg/A7D6F1BB18CD4CB48/xpcshell.sym"
-        ):
-            metadata_cache[api_params["Key"]] = api_params["Metadata"]
-            # ...pretend to actually upload it.
-            return {}
-
-        raise NotImplementedError((operation_name, api_params))
+    url = reverse("upload:upload_archive")
 
-    with botomock(mock_api_call), open(ZIP_FILE, "rb") as f:
-        response = client.post(url, {"file.zip": f}, HTTP_AUTH_TOKEN=token.key)
+    with open(ZIP_FILE, "rb") as fp:
+        response = client.post(url, {"file.zip": fp}, HTTP_AUTH_TOKEN=token.key)
         assert response.status_code == 201
 
         assert Upload.objects.all().count() == 1
-        assert FileUpload.objects.all().count() == 1
+        assert FileUpload.objects.all().count() == 2
 
-    # Upload the same file again. This time some of the S3 HeadObject
-    # operations should benefit from a cache.
-    with botomock(mock_api_call), open(ZIP_FILE, "rb") as f:
-        response = client.post(url, {"file.zip": f}, HTTP_AUTH_TOKEN=token.key)
+    # Upload the same file again. This time some of the S3 HeadObject operations should
+    # benefit from a cache.
+    #
+    # FIXME(willkg): we're not testing whether some of the lookups were from the cache
+    # or not
```

Comment:
The way caching is implemented makes it really hard to test against. When we rewrite the storage layer, we should architect it in a way that makes it possible (easy?) to test whether something was pulled from cache.